GDAXClient/Services/Products/ProductsService.cs

```diff
@@ -84,7 +84,35 @@ public class ProductsService : AbstractService
             return httpResponseMessage;
         }
 
-        public async Task<IEnumerable<object[]>> GetHistoricRatesAsync(ProductType productPair, DateTime start, DateTime end, int granularity)
+        /// <summary>
+        /// Get the Historic Rates and auto batch request when the date range is too large
+        /// </summary>
+        public async Task<IList<Candle>> GetHistoricRatesAsync(ProductType productPair, DateTime start, DateTime end, CandleGranularity granularity)
+        {
+            const int maxPeriods = 350; // From GDAX docs
+
+            var rc = new List<Candle>();
+
+            DateTime ? dEnd = end;
+            DateTime dStart;
+
+            var maxBatchPeriod = (int)granularity * maxPeriods;
+
+            // Work backwards, so the results are correctly ordered from latest to oldest
```

Comment:
can remove these comments. need to think about this logic a little bit more. but seems right to me.

Reply:
What isn't immediately obvious is that the DateTime values passed in are not directly used to set the start/end periods of the candle data coming back from GDAX

EG if you asked for 5 min data... with a start time of say 10:12 - the data coming back is for the GDAX 5 min candles at 10:15, 10:20, 10:25 etc. (IE not 10:12, 10:17, 10:22 etc.)

So we do need to use the datetime of the last candle of the batch coming back, as the End Time of the next batch... and then set the next batch start time backwards in time from that discovered date.